### PR TITLE
ECOM-6597/ECOM-6596 Management command for creating programs

### PR DESCRIPTION
**Pull request by MatthewPiatetsky**
Creates test partner, test program, and a program that looks like the wharton pc
@edx/ecommerce

### CODE REVIEW

**Comment by MatthewPiatetsky** [comment]
@edx/ecommerce based on the codecov looks like I need to write tests for this command. 
To be able to do this, I need to further understand the requirements.
Do we want this command to be run once, on every gocd deploy (except prod)?
Do we want the command to overwrite the test data if it already exists or do nothing in that case?

**Comment by mikedikan** [comment]
@MatthewPiatetsky I think at the outset, it won't be run by gocd.  Also, i think the overwrite idea makes the most sense to me, it's like hitting the reset button

**Comment by mikedikan** [comment]
jenkins run a11y

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+from course_discovery.apps.course_metadata.models import CourseRun, Program, ProgramType, Organization
+from course_discovery.apps.course_metadata.tests.factories import (
+    OrganizationFactory, CorporateEndorsementFactory, EndorsementFactory, ExpectedLearningItemFactory,
+    FAQFactory, JobOutlookItemFactory, ProgramFactory, CourseFactory, CourseRunFactory, PartnerFactory
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
```
Comment (mikedikan) [inline course_discovery/apps/course_metadata/management/commands/create_test_program.py:16]: Command seems like too generic of a name for this class - perhaps CreateTestDataCommand?

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        courses = [course_run.course for course_run in CourseRun.objects.filter(key__in=course_run_keys)]
+        wharton, _ = Organization.objects.get_or_create(key="Wharton")
+        ProgramFactory(
+            title="Test Digital Marketing Professional Certificate",
+            subtitle="Gain a competitive advantage and learn essential digital marketing skills and strategies in"
+            "this 4-course program from the top ranked Wharton School of the University of Pennsylvania.",
+            status=CourseRunStatus.Unpublished,
+            type=ProgramType.objects.get(name="Professional Certificate"),
+            partner=partner,
+            banner_image_url="https://www.edx.org/sites/default/files/prod-land-pg/banner/certificate_banner_1440x326.jpg",  # pylint: disable=line-too-long
```
Comment (mikedikan) [inline course_discovery/apps/course_metadata/management/commands/create_test_program.py:72]: Curious why you chose the pylint hints instead of breaking these up

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            program.first().delete()
+        logger.info('Creating Test Wharton Program')
+        course_run_keys = [
+            'course-v1:Wharton+CustomerCentricityx+3T2016',
+            'course-v1:Wharton+DigitalMarketing1.1x+3T2016',
+            'course-v1:Wharton+MarketingAnalytics101x+2T2016',
+            'course-v1:Wharton+SellingIdeas101x+3T2016'
+        ]
+        courses = [course_run.course for course_run in CourseRun.objects.filter(key__in=course_run_keys)]
+        wharton, _ = Organization.objects.get_or_create(key="Wharton")
```
Comment (mikedikan) [inline course_discovery/apps/course_metadata/management/commands/create_test_program.py:65]: what is the _ for?

File: course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+from course_discovery.apps.core.models import Partner
+from course_discovery.apps.course_metadata.models import Program
+from course_discovery.apps.course_metadata.tests.factories import PartnerFactory, ProgramFactory
+
+
+class CreateTestProgramCommandTests(TestCase):
+    def test_create_command(self):
+        call_command('create_test_program')
+        # Verify that these objects have been created
```
Comment (mikedikan) [inline course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py:12]: why are there no asserts in any of these tests?

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py (hunk 1 above)
Comment (MatthewPiatetsky) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:16]: All our management commands have this class, think it is standard

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py (hunk 3 above)
Comment (MatthewPiatetsky) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:65]: get_or_create return a tuple, the second value is a bool which indicates whether the object was created or not, which we don't care about.  
if you look at other uses of get_or_create in the catalog we use _ to assign that value that we don't care about

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py (hunk 2 above)
Comment (MatthewPiatetsky) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:72]: well these aren't lines of code just really long strings of text, it seemed like it would make it less rather than more readable to split each string into many many lines

File: course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py (hunk 4 above)
Comment (MatthewPiatetsky) [reply course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py:12]: I guess I could add asserts to check the value of specific data points on the program, but it seems to me like the main thing I care about is whether the objects actually get created or not? Most of the data is just some sort of random text from the factory

Comment (davec-edx) [reply course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py:12]: Are we also testing creation of a duplicate, verification that getting bad names, etc. will fail appropriately?  Ie, both positive and negative paths.

Comment (MatthewPiatetsky) [reply course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py:12]: This will overwrite an existing partner/program of the same name so you can't get a duplicate. Not sure what you mean by bad names

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py (hunk 1 above)
Comment (mikedikan) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:16]: 👌

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py (hunk 3 above)
Comment (mikedikan) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:65]: 👌

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py (hunk 2 above)
Comment (mikedikan) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:72]: that's fine, but on some lines you wrap the text, and on others you don't - can you make it consistent?

File: course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py (hunk 4 above)
Comment (mikedikan) [reply course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py:12]: 👌

File: course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        # If they have not they will throw a DoesNotExist error
+        Partner.objects.get(name='test-partner')
+        Program.objects.get(title='test-program')
+        Program.objects.get(title='Test Digital Marketing Professional Certificate')
+
+    def test_create_command_overwrite(self):
+        # Verify that these objects can be overwritten
+        PartnerFactory(name='test-partner')
```
Comment (mikedikan) [inline course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py]: I'm confused about this test - it's testing that by calling the command, no exception is thrown?  shouldn't we also test that you can still get the objects with the same title?

Comment (mikedikan) [reply course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py]: should we also test that it will overwrite data and reset to default (meaning, if you give title and subtitle, calling the command will reset the subtitle to the test default)?

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py (hunk 2 above)
Comment (MatthewPiatetsky) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:72]: made it consistent

File: course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py (hunk 5 above)
Comment (MatthewPiatetsky) [reply course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py]: changed test to set a value on the existing objects and check that the value is different after running the command

File: course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        PartnerFactory(name='test-partner', short_code='test')
+        ProgramFactory(title='test-program', subtitle='test')
+        ProgramFactory(title='Test Digital Marketing Professional Certificate', subtitle='test')
+        call_command('create_test_program')
+        partner = Partner.objects.get(name='test-partner')
+        test_program = Program.objects.get(title='test-program')
+        test_wharton_program = Program.objects.get(title='Test Digital Marketing Professional Certificate')
+        self.assertNotEqual(partner.short_code, 'test')
+        self.assertNotEqual(test_program.subtitle, 'test')
+        self.assertNotEqual(test_wharton_program.subtitle, 'test')
```
Comment (mikedikan) [inline course_discovery/apps/course_metadata/management/commands/tests/test_create_test_program.py:29]: 👌

**Review by mikedikan — APPROVED** [review]

**Review by bderusha — APPROVED** [review]
Overall I think this is fine.  My only concern is about extensibility.  Would all of the content be better contained in a YAML file that we manage and maintain?  Are there any attributes that we may want to programmatically create iterations of?

Not going to block this PR on any of that, but I would want to know what the plan is

File: course_discovery/apps/course_metadata/management/commands/create_test_program.py (hunk 1 above)
Comment (clintonb) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:16]: The base class is standard. The class name is not. Please rename the class to be more descriptive.

Comment (clintonb) [reply course_discovery/apps/course_metadata/management/commands/create_test_program.py:16]: Never mind. Disregard.

**Comment by MatthewPiatetsky** [comment]
@bderusha agreed (though personally I prefer json to yaml)
However, I don't think we currently have a plan for how we would extend this moving forward. 
I think it makes sense to just have something basic for now and then make it configurable once we know what we'll want to configure
